graphql2/graphqlapp/alert.go

```diff
@@ -177,6 +179,88 @@ func (q *Query) mergeFavorites(ctx context.Context, svcs []string) ([]string, er
 	return svcs, nil
 }
 
+// splitRangeByDuration splits the timeframe given between since and until by the duration provided.
+// Each segment is then transformed into an AlertDataPoint based on the given alerts.
+// The given alerts are required to be sorted by their CreatedAt field.
+func splitRangeByDuration(since, until time.Time, dur timeutil.ISODuration, alerts []alert.Alert) (result []graphql2.AlertDataPoint) {
+	if since.After(until) {
+		return result
+	}
+
+	if dur.IsZero() {
+		panic("duration must not be zero")
+	}
+
+	countAlertsUntil := func(ts time.Time) int {
+		var count int
+		for len(alerts) > 0 {
+			if !alerts[0].CreatedAt.Before(ts) {
+				break
+			}
+
+			count++
+			alerts = alerts[1:]
+		}
+		return count
+	}
+
+	// trim alerts before since
+	countAlertsUntil(since)
+
+	ts := since
+	for ts.Before(until) {
+		next := ts.AddDate(dur.Years, dur.Months, dur.Days).Add(dur.TimePart)
+		if next.After(until) {
+			next = until
+		}
+		result = append(result, graphql2.AlertDataPoint{
+			Timestamp:  ts,
+			AlertCount: countAlertsUntil(next),
+		})
+		ts = next
+	}
+
+	return result
+}
+
+func (q *Query) AlertMetrics(ctx context.Context, opts graphql2.AlertMetricsOptions) (result []graphql2.AlertDataPoint, err error) {
+	err = validate.Many(
+		validate.Range("ServiceIDs", len(opts.FilterByServiceID), 1, 1),
+		validate.TimeFrame("Timeframe", opts.Since, opts.Until, 1200*time.Hour),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	opts.Since = opts.Since.UTC()
+	opts.Until = opts.Until.UTC()
```

Comment:
I think the only issue is the timezone with the `splitRangeByDuration` helper -- can we move this to the function call, or maybe it better to have the function handle it internally?

We can comment on the helper that it performs the split in UTC -- if we ever need a timezone-aware version we can make one that takes a `time.Location` argument. If we fix it internally to the helper, the caller of the func doesn't need to be aware that if they forget to call `.UTC()` they could get unexpected results.